### ISSUE TITLE
[ENH] Sub-Optimal `FullAttention` Implementations Taken from TSLib

### ISSUE THREAD

**Issue by anasashb** [sktime/pytorch-forecasting#1990]
**Is your feature request related to a problem? Please describe.**
I've been observing the work being done for the `v2` release of `pytorch-forecasting`, including the implementation of the `TimeXer` model, taken from https://github.com/thuml/Time-Series-Library. 

Obviously, we all know that the TSLib is great for when it comes to releasing SOTA models to the public, but imo a lot of code in that original library is more or less outdated and un-optimized given what's possible in newer versions of torch. Why I'm highlighting this is that in TSLib torch is pinned at `1.7.1` (see [here](https://github.com/thuml/Time-Series-Library/blob/3846bfb81c8f3a86f154965819409d5a12a121c9/requirements.txt#L12C1-L12C13)), while in `pytorch-forecasting` minimum allowed version is at `torch==2.0.0`. 

Given this, imo there's no need to actually carry over the TSLib-s `torch.einsum()`-based attention computation to this project (see l66 of `layers/_attention/_full_attention.py` [here](https://github.com/sktime/pytorch-forecasting/blob/400b00f7c676ed869bcc026544743861112c1af7/pytorch_forecasting/layers/_attention/_full_attention.py#L66) and l67 of `/models/timexer/sub_modules.py` [here](https://github.com/sktime/pytorch-forecasting/blob/400b00f7c676ed869bcc026544743861112c1af7/pytorch_forecasting/models/timexer/sub_modules.py#L67) ).

The `.einsum()` computation is much more memory hungry and slower than for the `torch.nn.functional.scaled_dot_product_attention` now natively available in torch + the latter can use flash attn kernel if the user's GPU supports it. So I don't really see a reason to be using the slower implementation especially as *you don't need it*, because it seems you're not exposing the attention scores (see `output_attention=False` [here](https://github.com/sktime/pytorch-forecasting/blob/400b00f7c676ed869bcc026544743861112c1af7/pytorch_forecasting/models/timexer/_timexer.py#L275) in the `TimeXer` class) to the users anyways.

Besides the original TSLib modules have a lot of unused args kept in there for some specific models (e.g. the `tau` and `delta` in the `FullAttention` module are meant for the Non-Stationary Transformer specifically, and none of the more recent TSLib models use it). There's also a lot of other memory-related optimizations possible to the TSLib implementations anyways, such as using a single fused layer for Q,K,V projection and chunking them instead of separate projections and so on.

All of these things as of now seem to be carried over into your implementations of the models, so I think a lot could be improved ahead of the `v2` release so that this package can offer users *a major perf improvement over what's originally possible with TSLib models*.

**Describe the solution you'd like**

I'm not really requesting a solution from your side. If this is something that *you'd allow me to pick up* as an issue to refactor the TimeXer class accordingly, I'd be very happy. I've worked with TimeXer for my own use cases extensively and found major speedup and memory savings after refactoring the original outdated TSLib attention implementations.

**Describe alternatives you've considered**
Another alternative possible: If the reasoning for keeping `torch.einsum()` based attn implementation is that you want to have the attention scores accessible, we could also refactor the Attention classes to have two backends, one could be `.einsum()`-based, able to return the attention scores (as you have now) and another would be the `torch.nn.functional.scaled_dot_product_attention`-based, offering optimized perf but w/o the ability to return the attention scores.

**Additional context**
I've done some benchmarking tests between the two attention methods discussed here for my purposes and can share the results with you as well. If I can take this issue on, would include some test modules (+ maybe some kind of benchmarking notebook) to demonstrate the speedup to you as well.

Big fan of your work in general, and I'd be happy if I can contribute 🫡

**Comment by anasashb** [comment]
So just as an update for your side to have more details and to show that refactoring the TSLib models is a promising task:

I made a branch from my fork of `pytorch-forecasting` where I refactored the `FullAttention` class (and kept the existing version as `_FullAttention`) in the module meant for the v2 package version of `TimeXer` - [link to the module on my branch](https://github.com/anasashb/pytorch-forecasting/blob/feature/replace_einsum_attn/pytorch_forecasting/layers/_attention/_full_attention.py).

Then ran a small benchmark of the current version and the version I'd refactor the class into - [link to a demo notebook on my branch](https://github.com/anasashb/pytorch-forecasting/blob/feature/replace_einsum_attn/examples/attn_comp.ipynb).

I tested several cases - with CUDA and with CPU, as well as full precision (fp32), fp16 AMP, and bfp16 AMP - in forward and backward passes, checking for output equivalence, as well as gradient equivalence. 

Main comparisons are w.r.t computation speed and in case of CUDA, GPU memory used as well.

I'll copy some results here (`Einsum backend` refers to your existing implementation, and `PyTorch backend` refers to my proposed implementation):

```
INFO:test:================================================================================
INFO:test:CONFIG: batch=32, heads=8, seq=720, dim=256
INFO:test:Device: cuda, AMP: False (N/A)
INFO:test:================================================================================
INFO:test:EQUIVALENCE CHECK
INFO:test:Dtype: torch.float32
INFO:test:Tolerance: rtol=1e-05, atol=1e-05
INFO:test:All close: True
INFO:test:Max diff: 1.609325e-06
INFO:test:Mean diff: 4.146193e-08
INFO:test:================================================================================
INFO:test:FORWARD PASS
INFO:test:Einsum backend: 36.24 ms | 1827.50 MB
INFO:test:PyTorch backend: 11.32 ms | 353.75 MB
INFO:test:Speedup: 3.20x
INFO:test:Memory saved: 1473.75 MB (80.6%)
INFO:test:================================================================================
INFO:test:BACKWARD PASS
INFO:test:Einsum: 135.90 ms | 2423.75 MB
INFO:test:PyTorch: 40.89 ms | 538.22 MB
INFO:test:Speedup: 3.32x
INFO:test:Memory saved: 1885.53 MB
INFO:test:================================================================================
INFO:test:Gradient equivalence:
INFO:test:Query grad close: True
INFO:test:Key grad close: True
INFO:test:Value grad close: True
INFO:test:================================================================================
INFO:test:CONFIG: batch=32, heads=8, seq=720, dim=256
INFO:test:Device: cuda, AMP: True (torch.float16)
INFO:test:================================================================================
INFO:test:EQUIVALENCE CHECK
INFO:test:Dtype: torch.float16
INFO:test:Tolerance: rtol=0.01, atol=0.005
INFO:test:All close: True
INFO:test:Max diff: 1.464844e-03
INFO:test:Mean diff: 3.319979e-05
INFO:test:================================================================================
INFO:test:FORWARD PASS
INFO:test:Einsum backend: 28.34 ms | 1348.50 MB
INFO:test:PyTorch backend: 2.40 ms | 358.95 MB
INFO:test:Speedup: 11.80x
INFO:test:Memory saved: 989.55 MB (73.4%)
INFO:test:================================================================================
INFO:test:BACKWARD PASS
INFO:test:Einsum: 108.17 ms | 2129.50 MB
INFO:test:PyTorch: 5.92 ms | 481.96 MB
INFO:test:Speedup: 18.27x
INFO:test:Memory saved: 1647.55 MB
INFO:test:================================================================================
INFO:test:Gradient equivalence:
INFO:test:Query grad close: True
INFO:test:Key grad close: True
INFO:test:Value grad close: True
INFO:test:================================================================================
INFO:test:CONFIG: batch=32, heads=8, seq=720, dim=256
INFO:test:Device: cuda, AMP: True (torch.bfloat16)
INFO:test:================================================================================
INFO:test:EQUIVALENCE CHECK
INFO:test:Dtype: torch.bfloat16
INFO:test:Tolerance: rtol=0.02, atol=0.01
INFO:test:All close: True
INFO:test:Max diff: 1.367188e-02
INFO:test:Mean diff: 2.651215e-04
INFO:test:================================================================================
INFO:test:FORWARD PASS
INFO:test:Einsum backend: 38.37 ms | 1818.75 MB
INFO:test:PyTorch backend: 2.16 ms | 358.95 MB
INFO:test:Speedup: 17.79x
INFO:test:Memory saved: 1459.80 MB (80.3%)
INFO:test:================================================================================
INFO:test:BACKWARD PASS
INFO:test:Einsum: 143.34 ms | 2381.00 MB
INFO:test:PyTorch: 5.89 ms | 481.21 MB
INFO:test:Speedup: 24.33x
INFO:test:Memory saved: 1899.80 MB
INFO:test:================================================================================
INFO:test:Gradient equivalence:
INFO:test:Query grad close: True
INFO:test:Key grad close: True
INFO:test:Value grad close: True
```

As you can see refactoring the Attention Layer does actually promise improvements. 

*If I get the greenlight from you on this feature, I'd happily do a refactoring and make a draft PR down the line*

- - -

Q, K, V Tensor shapes for the computations are: batch size of 32, context length of 720, number of attention heads 8, and model / embedding dimension of 256

My device / env specs for the benchmarking test are:
- Python 3.11.2
- `torch==2.9.0`
- GPU: NVIDIA RTX 500 Ada Generation Laptop GPU
- CUDA Version: 12.8

**Comment by phoeenniixx** [comment]
Sorry @anasashb, for the late reply. About your issue:

> If I get the greenlight from you on this feature, I'd happily do a refactoring and make a draft PR down the line

Thanks  for the proposal!
I really liked your idea and the results you shared! 
Can you please raise a PR here so the core devs can look at it more closely?
I just have a few questions:
- Are the shared results from [v1](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/_timexer.py) or [v2](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/_timexer_v2.py) implementation of `timexer`?
- Does your feature request also change the model implementation along with attention mechanism? If yes, then what are these changes?
- Are you planning to add this feat to v1 or v2? Is your feature request mainly just for `layers` module? `layers` module is for now very much independent of v1/v2 refactoring. so, we might need to think about these questions.
- For v1, have you looked at [sub_modules](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/sub_modules.py), I have not looked at your code, so I just want you to make sure, if your feature request is not present in the `sub_modules` as well, which are used by v1.

**Comment by phoeenniixx** [comment]
Just a heads-up:
- Please add the summary of changes and these results in the PR description, it will be really helpful for the reviewers. If you want, you can also add an example nb comparing both the attention mechanisms.
-  Also, please dont remove the current `.einsum()` based implementation, rather, you can add a param where you can have the user select which kind of implementation they want, so that the people already using `.einsum()` (who dont want to switch to `scaled_dot_product_attention` based implementation) can also use `timexer`. 
- If you want, you can also add this new implementation to `sub_modules` as well which `timexer-v1` uses.

Currently, we have 2 versions- v1 and v2. This will remain the case until v2's API becomes stable and we deprecate v1, so for now, we need to think about both the versions whenever we make any changes to the implementations to the models which exist in both the versions 

I know this is a lot to navigate with the v1/v2 split, so please don't hesitate to ask for clarification on any of these points. We're happy to guide you on the best way to integrate this.

**Comment by anasashb** [comment]
@phoeenniixx Thanks for the response.

To answer your questions:

-  > Are the shared results from [v1](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/_timexer.py) or [v2](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/_timexer_v2.py) implementation of timexer?

The results I shared are from v2 `timexer`, but technically it does not matter because I just benchmarked the attention computation speed and memory consumption in the attention module directly. 

So, the changes I made was to the `FullAttention` class for the v2 under [layers/attention/_full_attention](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/layers/_attention/_full_attention.py#L30). From what I saw, the implementation of the `FullAttention` there is identical to the one v1 uses from [sub_modules](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/sub_modules.py).

- - -

- > Does your feature request also change the model implementation along with attention mechanism? If yes, then what are these changes?

Initially I was thinking to suggest other changes too within the model - such as switching to a fused projection in the `AttentionLayer`, instead of the individual Q, K, V projections being done currectly:

```python
        queries = self.query_projection(queries).view(B, L, H, -1)
        keys = self.key_projection(keys).view(B, S, H, -1)
        values = self.value_projection(values).view(B, S, H, -1)
```
And to possible get rid of redundant arguments that are not used anywhere such as `tau` and `delta` here in `AttentionLayer`'s forward (these are just carried over from `tslib`):

```python

def forward(self, queries, keys, values, attn_mask, tau=None, delta=None):
```

*But*, given that this would be my first PR to your package, I think it is better to just stick with a single change, the attention computation refactoring for now.

- - -

-  > For v1, have you looked at [sub_modules](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/timexer/sub_modules.py), I have not looked at your code, so I just want you to make sure, if your feature request is not present in the sub_modules as well, which are used by v1. ... If you want, you can also add this new implementation to sub_modules as well which timexer-v1 uses. 

 
About the several points you made about v1/v2: I've had a good look at the package since posting this issue and I agree. I'll implement this feature for both the v1 and the v2 versions, yes. 

- - -

- > Also, please dont remove the current .einsum() based implementation, rather, you can add a param where you can have the user select which kind of implementation they want, so that the people already using .einsum() (who dont want to switch to scaled_dot_product_attention based implementation) can also use timexer.

Agreed, that is the best option + I'll keep the `.einsum()` as default to make sure the update would be completely backwards compatible for anyone that may already be using the model  from the current version of the package.

- - -

FYI, I'll be picking up the work for this PR next week

**Comment by phoeenniixx** [comment]
Thanks @anasashb for the replies! we'll be looking forward to your PR!
If you need any help, you can also ping us on discord: https://discord.gg/scgEMN8e

**Comment by anasashb** [comment]
Hi @phoeenniixx, I've opened the PR:
- https://github.com/sktime/pytorch-forecasting/pull/1997

**Comment by phoeenniixx** [comment]
Thank you @anasashb! 
I will look into it in few days... I am a little busy with college this week
